### PR TITLE
[refactor]refactor lazy materialized

### LINKED ISSUE #8794: [Enhancement] Refactor lazy materialization
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

For the delayed materialization logic of the current storage layer, this is still some optimization space.
1 The decision rule for lazy materialization is not very clear, not so easy to understand.
2 Only the read and calculation overhead of the string type is considered, but the date type and decimal type are ignored. Lazy Materialization can do better in datetime/date/decimal case.

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wangbo**
# Proposed changes

Issue Number: close #8794

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by wangbo** [comment]
This PR has passed SSB test，query result is correct and no performance rollback.
I'll test more cast later.

**Comment by wangbo** [comment]
I've done a fairly thorough test. In including query with delete condition/ssb diy sql/tpch/bitmap/aggregation.
It is ready to merge.

File: be/src/olap/rowset/segment_v2/segment_iterator.h [hunk 1]
```diff
@@ -146,6 +146,8 @@ class SegmentIterator : public RowwiseIterator {
     // remember the rowids we've read for the current row block.
     // could be a local variable of next_batch(), kept here to reuse vector memory
     std::vector<rowid_t> _block_rowids;
+    bool _is_need_vec_eval = false;
+    bool _is_need_short_eval = false;
 
     // fields for vectorization execution
     bool _is_all_column_basic_type;
```
Comment (BiteTheDDDDt) [inline be/src/olap/rowset/segment_v2/segment_iterator.h:153]: If `_is_all_column_basic_type` is not used, the definition can be deleted.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 2]
```diff
@@ -625,13 +647,14 @@ void SegmentIterator::_vec_init_lazy_materialization() {
             _is_pred_column[cid] = true;
             pred_column_ids.insert(cid);
 
+            // Step1: check pred using short eval or vec eval
             if (type == OLAP_FIELD_TYPE_VARCHAR || type == OLAP_FIELD_TYPE_CHAR ||
                 type == OLAP_FIELD_TYPE_STRING || predicate->type() == PredicateType::BF ||
                 predicate->type() == PredicateType::IN_LIST ||
-                predicate->type() == PredicateType::NO_IN_LIST) {
+                predicate->type() == PredicateType::NO_IN_LIST ||
+                type == OLAP_FIELD_TYPE_DATE || type ==OLAP_FIELD_TYPE_DECIMAL) {
```
Comment (BiteTheDDDDt) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:655]: need format

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 3]
```diff
@@ -646,75 +669,82 @@ void SegmentIterator::_vec_init_lazy_materialization() {
...
+                FieldType type = _schema.column(cid)->type();
+                
+                // todo(wb) maybe we can make read char type faster
+                // todo(wb) support map/array type
+                // todo(wb) consider multiple integer columns cost, such as 1000 columns, maybe lazy materialization faster
+                if (!_lazy_materialization_read && 
+                        (_is_need_vec_eval || _is_need_short_eval) && // only when pred exists, we need to consider lazy materialization
+                        (type == OLAP_FIELD_TYPE_HLL || type == OLAP_FIELD_TYPE_OBJECT || 
+                        type == OLAP_FIELD_TYPE_VARCHAR || type == OLAP_FIELD_TYPE_CHAR || type == OLAP_FIELD_TYPE_STRING ||
+                        type == OLAP_FIELD_TYPE_BOOL || type == OLAP_FIELD_TYPE_DATE || type == OLAP_FIELD_TYPE_DATETIME ||
```
Comment (BiteTheDDDDt) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:709]: `OLAP_FIELD_TYPE_BOOL` is slow type too?

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 4]
```diff
@@ -937,48 +969,48 @@ Status SegmentIterator::next_batch(vectorized::Block* block) {
...
-                                                      selected_size));
-        }
+        // 4.2 get union of short_cir_pred and vec_pred
+        std::set<ColumnId> pred_column_ids;
+        pred_column_ids.insert(_short_cir_pred_column_ids.begin(), _short_cir_pred_column_ids.end());
+        pred_column_ids.insert(_vec_pred_column_ids.begin(), _vec_pred_column_ids.end());
+
+        // 4.3 output short circuit and predicate column
+        RETURN_IF_ERROR(_output_column_by_sel_idx(block, pred_column_ids, sel_rowid_idx,
+                                                    selected_size));
```
Comment (zenoyang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1013]: Executed here, it means that `_lazy_materialization_read` is true, then `pred_column_ids` can be replaced by `_first_read_column_ids`, and the above three lines can be deleted.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 3 above)
Comment (wangbo) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:709]: I didn't do performance test.
But bool is read from a runlenDecoder, it means bool can only be read row by row, can not in batch.

File: be/src/olap/rowset/segment_v2/segment_iterator.h (hunk 1 above)
Comment (wangbo) [reply be/src/olap/rowset/segment_v2/segment_iterator.h:153]: 👌🏻

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 4 above)
Comment (wangbo) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:1013]: 👌🏻

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 2 above)
Comment (wangbo) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:655]: 👌🏻

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**